### PR TITLE
gh-105539: Explict resource management for connection objects in sqlite3 tests

### LINKED ISSUE #105539: Emit resource warning if sqlite3 fails to close the database
_See https://github.com/python/cpython/issues/103837#issuecomment-1541808307 and https://github.com/python/cpython/issues/103837#issuecomment-1543936079_:

EAA:
> I still think we should consider changing the underlying API so we can emit a resource warning if close failed.

VS:
> IMO sqlite3 should be modified to emit a ResourceWarning.
>
> Currently, no ResourceWarning is emitted when a connection is not closed explicitly:

### Linked PRs
* gh-108015
* gh-108017
* gh-108360
* gh-108565

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-105539

### CODE REVIEW

**Comment by erlend-aasland** [comment]
This PR either uses the `memory_database()` resource management helper from `Lib/test/test_sqlite3/test_dbapi.py`, or introduces explicit resource management via `setUp()`/`tearDown()` test case methods.

File: Lib/test/test_sqlite3/test_dbapi.py [hunk 1]
```diff
@@ -1659,97 +1668,87 @@ def test_cursor_executescript_too_large_script(self):
...
+        self.con = sqlite.connect(":memory:")
+        self.cur = self.con.cursor()
+        self.con.close()
+
     def test_closed_con_cursor(self):
-        con = sqlite.connect(":memory:")
-        con.close()
-        with self.assertRaises(sqlite.ProgrammingError):
-            cur = con.cursor()
+        with self.assertRaisesRegex(sqlite.ProgrammingError, self.REGEX):
```
Comment (vstinner) [inline Lib/test/test_sqlite3/test_dbapi.py]: You may move ``self.assertRaisesRegex(sqlite.ProgrammingError, self.REGEX)`` in a helper function.

**Review by vstinner — COMMENTED** [review]
LGTM.

I would suggest moving memory_database() to a new ``utils.py`` file. For me, it's surprising that tests import other tests. See for example ``Lib/test/test_asyncio/utils.py``.

**Review by vstinner — APPROVED** [review]

**Comment by vstinner** [comment]
I don't know the DB API, but I'm surprised that ``with connect() as db: ...`` doesn't close the database :-(

File: Lib/test/test_sqlite3/test_hooks.py [hunk 2]
```diff
@@ -203,31 +207,35 @@ def progress():
...
+        self.con.set_progress_handler(bad_progress, 1)
         with self.assertRaises(sqlite.OperationalError):
-            con.execute("""
+            self.con.execute("""
                 create table foo(a, b)
                 """)
 
 
 class TraceCallbackTests(unittest.TestCase):
+    def setUp(self):
```
Comment (sobolevn) [inline Lib/test/test_sqlite3/test_hooks.py]: This is also a pattern: maybe a test class mixin?

File: Lib/test/test_sqlite3/test_userfunctions.py [hunk 3]
```diff
@@ -405,19 +405,19 @@ def test_func_deterministic_keyword_only(self):
     def test_function_destructor_via_gc(self):
         # See bpo-44304: The destructor of the user function can
         # crash if is called without the GIL from the gc functions
-        dest = sqlite.connect(':memory:')
-        def md5sum(t):
-            return
+        with memory_database() as dest:
+            def md5sum(t):
```
Comment (sobolevn) [inline Lib/test/test_sqlite3/test_userfunctions.py]: Function should be outside of the context manager, I guess.

File: Lib/test/test_sqlite3/test_regression.py [hunk 4]
```diff
@@ -295,29 +287,31 @@ def test_convert_timestamp_microsecond_padding(self):
...
+                datetime.datetime(2012, 4, 4, 15, 6, 0, 123456),
+            ])
 
     def test_invalid_isolation_level_type(self):
         # isolation level is a string, not an integer
-        self.assertRaises(TypeError,
-                          sqlite.connect, ":memory:", isolation_level=123)
+        regex = "isolation_level must be str or None"
+        with self.assertRaisesRegex(TypeError, regex):
+            with memory_database(isolation_level=123):
```
Comment (sobolevn) [inline Lib/test/test_sqlite3/test_regression.py]: `memory_database(isolation_level=123).__enter__()`?

This way we are more specific :)

File: Lib/test/test_sqlite3/test_regression.py [hunk 5]
```diff
@@ -238,7 +231,8 @@ def test_auto_commit(self):
         2.5.3 introduced a regression so that these could no longer
         be created.
         """
-        con = sqlite.connect(":memory:", isolation_level=None)
+        with memory_database(isolation_level=None) as con:
```
Comment (sobolevn) [inline Lib/test/test_sqlite3/test_regression.py:231]: This does not feel complete: can we assert something here? 🤔

File: Lib/test/test_sqlite3/test_regression.py [hunk 6]
```diff
@@ -45,28 +45,24 @@ def test_pragma_user_version(self):
 
     def test_pragma_schema_version(self):
         # This still crashed pysqlite <= 2.2.1
-        con = sqlite.connect(":memory:", detect_types=sqlite.PARSE_COLNAMES)
-        try:
+        with memory_database(detect_types=sqlite.PARSE_COLNAMES) as con:
             cur = self.con.cursor()
```
Comment (sobolevn) [inline Lib/test/test_sqlite3/test_regression.py:46]: `cur` is not closed now, I think that `con.close()` will close it, just want to double check.

File: Lib/test/test_sqlite3/test_hooks.py (hunk 2 above)
Comment (erlend-aasland) [reply Lib/test/test_sqlite3/test_hooks.py]: Yeah, I thought about that myself. Let's do that now that we have the opportunity.

File: Lib/test/test_sqlite3/test_userfunctions.py (hunk 3 above)
Comment (erlend-aasland) [reply Lib/test/test_sqlite3/test_userfunctions.py]: Good catch!

File: Lib/test/test_sqlite3/test_regression.py (hunk 5 above)
Comment (erlend-aasland) [reply Lib/test/test_sqlite3/test_regression.py:231]: That test has been like that for years. I'm not sure it is really helpful anymore.

File: Lib/test/test_sqlite3/test_regression.py (hunk 6 above)
Comment (erlend-aasland) [reply Lib/test/test_sqlite3/test_regression.py:46]: Yeah, the connection object implicitly closes all open cursors in its `tp_dealloc`

File: Lib/test/test_sqlite3/test_dbapi.py (hunk 1 above)
Comment (erlend-aasland) [reply Lib/test/test_sqlite3/test_dbapi.py]: Good suggestions; I'll do those before landing.

**Comment by erlend-aasland** [comment]
I think all your remarks are addressed now; PTAL :)

**Review by vstinner — APPROVED** [review]
LGTM.

If you want, you can replace ``from test.test_sqlite3.util import`` with ``from .util import``.

**Comment by erlend-aasland** [comment]
Thanks for the review, Victor and Nikita!

File: Lib/test/test_sqlite3/test_hooks.py (hunk 2 above)
Comment (erlend-aasland) [reply Lib/test/test_sqlite3/test_hooks.py]: You happy with the mixin?

**Review by sobolevn — APPROVED** [review]
My first `sqlite3` PR review ✅

**Pull request merged**